Log out auth subject on Bolt session halt

henriknyman opened the pull request:
This is to make sure we don't leave auth info cached

henriknyman committed:
Logout auth subject on Bolt session halt

pontusmelke commented on community/bolt/src/main/java/org/neo4j/bolt/security/auth/BasicAuthentication.java:
```diff
@@ -70,9 +70,19 @@ public AuthenticationResult authenticate( Map<String,Object> authToken ) throws
+    @Override
+    public void logout( AccessMode accessMode )
+    {
+        if ( accessMode instanceof AuthSubject )
```
It feels like somewhat of a design flaw that we need to do these casts. Could we have something along the line of

```
public interface Authentication<T extends AccessMode>
{
...
    public void logout( T accessMode );
}
```

and then have `BasicAuthentication implements Authentication<AuthSubject>`

henriknyman replied:
That sounds nice, but I think currently BasicAuthentication have to deal with AccessMode.Static coming in (the session state is initialized with AccessMode.Static.NONE). 
I don't really like the looks of logout taking AccessMode, so another alternative I was considering is pulling up the AuthSubject into the session state instead of AccessMode.. What do you think?

henriknyman commented on community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/SessionStateMachine.java:
```diff
@@ -455,6 +455,10 @@ public State halt( SessionStateMachine ctx )
             }
+            if ( ctx.accessMode != null )
+            {
+                ctx.spi.logout( ctx.accessMode );
```
Maybe this should be wrapped in a try-catch as well

henriknyman committed:
Replace AccessMode with AuthSubject in Bolt

Refactoring to pull AuthSubject up the stack
- Store AuthSubject instead of AccessMode in the session state
- Initialize with AuthSubject.ANONYMOUS that has no access rights (for the
un-authenticated state)
- Add constant AuthSubject.AUTH_DISABLED with full access rights

craigtaverner commented:
I like the direction this has taken, removing AccessMode and basing it all on AuthSubject. Much cleaner.
